### PR TITLE
vendor/go-qemu: fixing possible mixed up response with events

### PR DESCRIPTION
**Pull request by rouming**
This PR targets nasty unexpected "" (empty) response on "query-status" QMP request (which is considered as request which should work in all QEMU states). Empty response is considered as fatal and EVE stops QEMU process. 

The PR targets vendor/go-qemu, I also upstreamed the fix https://github.com/digitalocean/go-qemu/pull/210, but I'm not sure when it is accepted and merged, so changing vendor code directly in EVE.

Patch drains possible events on qmp socket until true "qmp_capabilities" response is received.

This targets a nasty and rare problem in "Connect(), Run()" sequence, when asynchronous QEMU "event" just after the "qmp_capabilities" request is accepted as response and further any QMP request gets completed with response from the preceding "qmp_capabilities" response.

The following QMP protocol sequence is possible:

  1 read  {"QMP": {"version": {"qemu": {"micro": 0, "minor": 1, "major": 5}, ...}"
  2 write {"execute":"qmp_capabilities"}
  3 read  {"timestamp": {"seconds": xxx, "microseconds": xxx}, "event": "..."}
  4 read  {"return": {}}

the 3. is unexpected by the current Connect() implementation and "event" is considered as a proper response on "qmp_capabilities", in other turn 4. is read in the go.mos.listen() and immediately pushed to the stream channel, so any further QMP command (Run() call) will be immediately completed by an empty response from line 4.

The described problem of unexpected empty response line was observed on this code qmp.SocketMonitor sequence:

   Connect()
   Run('{"execute":"query-status"}') <<< Returns empty response
   Disconnect()

The problem is very rare and was observed ~5 times over a fairly long period of time (several months), which corresponds to nature of the described rare protocol race.

The current patch was tested on modified QEMU, where an aritifical sleep() was introduced in the qmp_marshal_qmp_capabilities() call just right after the qmp_qmp_capabilities() was invoked, so all further events can be accepted by the QMP socket:

```
 --- qapi/qapi-commands-control.c        2023-11-08 08:55:16.209007741 +0100
 +++ qapi/qapi-commands-control.c.orig   2023-11-08 08:55:13.929005997 +0100
 @@ -42,10 +42,6 @@

      qmp_qmp_capabilities(arg.has_enable, arg.enable, &err);

 +    printf(">>> BEFORE SLEEP 10\n");
 +    sleep(10);
 +    printf(">>> AFTER SLEEP 10\n");
 +
      error_propagate(errp, err);

  out:
```

Any QMP event can be freely received by the QMP client, while the execution flow of the qmp_marshal_qmp_capabilities() was interrupted or scheduled out.

The fix of the described is simple: read from the QMP socket until response is received and drop all possible events.

### CODE REVIEW

**Comment by OhmSpectator** [comment]
Wow. I still do not fully understand the code, but I get the idea.
I'm curious: how did you come to this fix?

**Review by OhmSpectator — COMMENTED** [review]
Wow. I still do not fully understand the code, but I get the idea.
I'm curious: how did you come to this fix?

**Comment by rouming** [comment]
> Wow. I still do not fully understand the code, but I get the idea. I'm curious: how did you come to this fix?

Stared into the abyss for a long time :)

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (OhmSpectator) [inline pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:140]: Could there be a risk of an infinite loop if non-capability events keep coming in without any actual progress?

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (OhmSpectator) [inline pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:149]: In the case of receiving JSON data without an Event field, should we consider additional validation to ensure the data is still relevant to our needs, or do we only expect event-related JSON in this asynchronous flow?

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (OhmSpectator) [inline pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:160]: Are we potentially overlooking important information by discarding events before the capability response?

**Comment by OhmSpectator** [comment]
I'm not an expert, but I'm afraid of some potential problems with the vendoring part...
The version of the lib used by pillar is mentioned here: https://github.com/lf-edge/eve/blob/171eb7fd46ed1e4e388d85eb8720625fc67c3794/pkg/pillar/go.mod#L12
Despite the fact we committed the change in the vendored version, the version mentioned in `go.mod` and `go.sum` can somehow affect the build/rebuild process...
Or is it a common way in our project to commit this kind of changes?

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (rouming) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:140]: The whole implementation of this library does not imply the timeout on response. So we hang here and on the Run() (commands invocation path) as well if QEMU qmp loop hangs.

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (rouming) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:149]: all possible validation is below in the same if branch

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (rouming) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:160]: what important information?

Comment (rene) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:160]: Doesn't make sense to at least log discarded events?

**Comment by rouming** [comment]
> Or is it a common way in our project to commit this kind of changes?

It's not. Once merged upstream we update the sha. If not merged upstream - we need to pull go-qemu repo with patches to our own repo. But for now this is is the easiest way.

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (OhmSpectator) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:149]: There, we assume that we wait for any reply that is a valid JSON and does not contain Error, right? Okay, the rest of the checks are up to the caller, I guess.

File: pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go
Comment (OhmSpectator) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:160]: I don't know... Is it possible that another entity is waiting for the events that we are dropping? I guess the answer is no, but it's better to ask :)

Comment (rouming) [reply pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go:160]: All events will be received on the listener QMP socket we have (see the `qmpEventHandler()`),  Here in the `Connect()`  we are not interested in events until a valid response is received.

**Comment by OhmSpectator** [comment]
@rouming I've found another patch that we already apply to the same file:

https://github.com/lf-edge/eve/blob/987d94a812ab9d13be344c38e84602a341d6afdd/pkg/pillar/pillar-patches/003-go-qemu-qmp-hung-listener.patch

Maybe it would make sense to do the same? Just to be consistent... Or... I don't know.

**Comment by milan-zededa** [comment]
> > Or is it a common way in our project to commit this kind of changes?
> 
> It's not. Once merged upstream we update the sha. If not merged upstream - we need to pull go-qemu repo with patches to our own repo. But for now this is is the easiest way.

Note that your change will be reverted the moment somebody calls `go mod vendor` in pillar. It is better to switch to your fork of go-qemu for the time being.

**Comment by rouming** [comment]
> > > Or is it a common way in our project to commit this kind of changes?
> > 
> > 
> > It's not. Once merged upstream we update the sha. If not merged upstream - we need to pull go-qemu repo with patches to our own repo. But for now this is is the easiest way.
> 
> Note that your change will be reverted the moment somebody calls `go mod vendor` in pillar. It is better to switch to your fork of go-qemu for the time being.

@milan-zededa  that will obliged me to keep the repo copy permanently and probably maintain it. and what if I'm laid off? I can't risk EVE.

**Review by milan-zededa — APPROVED** [review]
Run tests only (not approving direct change in vendor :))

**Review by eriknordmark — APPROVED** [review]
We cant merge this any go get update might update the vendor dir and clobber the change. But kicking off tests so see what happens.

**Comment by rouming** [comment]
> We cant merge this any go get update might update the vendor dir and clobber the change. But kicking off tests so see what happens.

Yeah, that's expected. I mainly pushed that as-is for discussion and to trigger tests. I will rework as pillar-patches.

**Comment by rouming** [comment]
@milan-zededa @eriknordmark Now the patch is in the patch. Please take a look.

This reminds me an old exhibit meme, for example: https://imgflip.com/i/85goo7

File: pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch
Comment (OhmSpectator) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: I've checked it and am still confused...
The patch tool is called from a copy of a pkg/pillar dir. In this case, I don't understand how the original file is found by the tool. It should look for `pkg/pillar/vendor/github.com/digitalocean/go-qemu/qmp/socket.go`, and it's not the location of the file in the working dir.

https://github.com/lf-edge/eve/blob/94fab914ac46f4531e734bb3afae4d4942cd015a/pkg/pillar/Dockerfile#L77-L94

Ok, heck it. If you have tested it and see both patches are applied successfully - let it be.

Comment (rouming) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: Pull the pr, make a compilation error in the patch and try to run make pkg/pillar if you have doubts.

Comment (OhmSpectator) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: Ok, I pulled the PR, modified the pillar Makefile to build only up to the "build" target, built the container, ran it and evaluated the resulting files. Both patches are applied, and everything looks fine.

But it's an awful workflow. The patch files that are applied only in temporary containers should go one day...

Comment (rouming) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: What stops you from making a fork of the go-qemu under the zededa umbrella, push two commits there and update go.mod? Sounds like 5 lines change and huge improvement.

Comment (OhmSpectator) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: Because it will check the code itself and not the way how it's built in the scope of our build system.

Comment (rouming) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: We probably talk about different things. The simple solution to this is to make our own repo of go-qemu under zededa, push 2 commits there, and change the vendor link in the go.mod, also not forgetting to remove pillar-patches as a folder and from dockerfile. This exactly changes the build system.

Comment (rouming) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: @OhmSpectator all in your hands: https://github.com/zededa/go-qemu please proceed further.

Comment (OhmSpectator) [reply pkg/pillar/pillar-patches/004-vendor-go-qemu-fixing-possible-mixed-up-response-wit.patch:71]: I'm not saying we should do it right now.
The most frustrating part of it - you don't see the resultsing code that is really build. For that I added a new targets into Makefile and created the PR. 
I can leave with that for some time.
And replacing the entire repo with our fork - I don't think it's 5 minutes action.

**Review by eriknordmark — APPROVED** [review]
LGTM

**Pull request merged**
